crates/autopilot/src/domain/settlement/trade/mod.rs

```diff
@@ -0,0 +1,184 @@
+use {
+    super::{transaction, transaction::Prices},
+    crate::domain::{
+        self,
+        auction::{self, order},
+        eth,
+        fee,
+    },
+    bigdecimal::Zero,
+};
+
+mod math;
+
+/// Trade type evaluated in a context of an Auction.
+#[derive(Clone, Debug)]
+pub enum Trade {
+    /// A regular user order that exist in the associated Auction.
+    Fulfillment(Fulfillment),
+    // JIT trades are not part of the orderbook and are created by solvers at the time of
+    // settlement.
+    // Note that user orders can also be classified as JIT orders if they are settled outside of
+    // the Auction.
+    Jit(Jit),
+}
+
+impl Trade {
+    /// UID of the order that was settled in this trade.
+    pub fn uid(&self) -> &domain::OrderUid {
+        match self {
+            Self::Fulfillment(trade) => &trade.uid,
+            Self::Jit(trade) => &trade.uid,
+        }
+    }
+
+    /// Return JIT order if it's a JIT order.
+    pub fn as_jit(&self) -> Option<&Jit> {
+        match self {
+            Self::Fulfillment(_) => None,
+            Self::Jit(trade) => Some(trade),
+        }
+    }
+
+    /// CIP38 score defined as surplus + protocol fee
+    pub fn score(&self, auction: &super::Auction) -> Result<eth::Ether, math::Error> {
+        match self {
+            Self::Fulfillment(trade) => math::Trade::from(trade.clone()).score(auction),
+            Self::Jit(trade) => math::Trade::from(trade.clone()).score(auction),
+        }
+    }
+
+    /// Surplus of a trade.
+    pub fn surplus_in_ether(&self, prices: &auction::Prices) -> Result<eth::Ether, math::Error> {
+        match self {
+            Self::Fulfillment(trade) => math::Trade::from(trade.clone()).surplus_in_ether(prices),
+            Self::Jit(trade) => {
+                if trade.surplus_capturing {
+                    math::Trade::from(trade.clone()).surplus_in_ether(prices)
+                } else {
+                    // JIT orders that are not surplus capturing have zero
+                    // surplus, even if they settled at a better price than
+                    // limit price.
+                    Ok(eth::Ether::zero())
+                }
+            }
+        }
+    }
+
+    /// Total fee taken for the trade.
+    pub fn fee_in_ether(&self, prices: &auction::Prices) -> Result<eth::Ether, math::Error> {
+        match self {
+            Self::Fulfillment(trade) => math::Trade::from(trade.clone()).fee_in_ether(prices),
+            Self::Jit(trade) => math::Trade::from(trade.clone()).fee_in_ether(prices),
+        }
+    }
+
+    /// Total fee (protocol fee + network fee). Equal to a surplus difference
+    /// before and after applying the fees.
+    pub fn fee_in_sell_token(&self) -> Result<eth::SellTokenAmount, math::Error> {
+        match self {
+            Self::Fulfillment(trade) => math::Trade::from(trade.clone()).fee_in_sell_token(),
+            Self::Jit(trade) => math::Trade::from(trade.clone()).fee_in_sell_token(),
+        }
```

Comment:
nit: it might make more sense to implement From<Trade> instead of having these two match arms everywhere.